equinix provider: initial implementation

tdomnesc opened the pull request:
as commented in the provider the packngo package requires the equinix tocken env var PACKET_AUTH_TOKEN

derekhiggins commented on config/samples/ofcir_v1_cipool.yaml:
```diff
@@ -93,3 +93,32 @@ stringData:
+stringData:
+  config: |
+    {
+      "projectid": "f2c6f0a6-4673-464b-878d-cce8dc8d0648",
```
lets not put a real project ID in here

derekhiggins commented on pkg/providers/equinix.go:
```diff
@@ -0,0 +1,121 @@
+
+func EquinixProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := equinixProviderConfig{
+		ProjectID: "f2c6f0a6-4673-464b-878d-cce8dc8d0648",
```
same here

derekhiggins commented on pkg/providers/equinix.go:
```diff
@@ -0,0 +1,121 @@
+	resource := Resource{}
+
+	uniqueId := strings.Replace(uuid.New().String(), "-", "", -1)
+	resourceName := fmt.Sprintf("%s-%s", hostPrefix, uniqueId)
```
Instead of using a unique ID here we could base the  "resourceName" on the CIR somehow so that it consistent.
Then before creating the device, check if a device with the same name already exists. If it does we could either
use it, delete it or just fail.

This would ensure we only ever create a single device per CIR

derekhiggins commented on pkg/providers/equinix.go:
```diff
@@ -0,0 +1,121 @@
+	if err != nil {
+		return resource, err
+	}
+
```
we could also test if ssh is available here, I've a branch to do something similar here, but no a PR yet as its waiting for the auth PR to merge
https://github.com/openshift-eng/ofcir/commit/ca441958144c665708e3d4805f2dab93dd23a090

tdomnesc replied:
I will do this once your utils package merges

andfasano commented on config/samples/ofcir_v1_cipool.yaml:
```diff
@@ -93,3 +93,32 @@ stringData:
+stringData:
+  config: |
+    {
+      "projectid": "replace-with-project-id",
```
What about the token? I was expecting it to be defined here

tdomnesc replied:
https://github.com/packethost/packngo requires the token to be exported as env var PACKET_AUTH_TOKEN. Should I export it from the provider with os.Setenv?

andfasano replied:
I think it could be possible also to use `packngo.NewClientWithAuth()` [function](https://github.com/packethost/packngo/blob/1d7ae3f7dbacd747ad8573c7daa246d0f8b2a9a8/packngo.go#L350-L353), which accepts an api token

andfasano commented on pkg/providers/equinix.go:
```diff
@@ -0,0 +1,121 @@
+		}
+	}
+
+	client, err := packngo.NewClient()
```
This gets created @ every reconcile loop, wouldn't be better to have a singleton?

derekhiggins commented on pkg/providers/equinix.go:
```diff
@@ -0,0 +1,142 @@
+			count++
+		}
+	}
+	if count >= poolSize {
```
we may have a problem here if there are multiple equinix pools all sharing the same "hostPrefix" but pool specific poolSize. I'm ok with fixing this later as it likely wont be a problem right away,

tdomnesc replied:
Not sure why we would have multiple pools in the same equinix project, but I understand the problem.

andfasano replied:
One reasonable case would be also to have different pools for different equinix configurations, for example the family gen.

derekhiggins commented:
1 comment inline but looks good to me
some of the commits are fixing up previous ones, do you want to squash them together before merging?

tdomnesc commented:
> 1 comment inline but looks good to me some of the commits are fixing up previous ones, do you want to squash them together before merging?

Rebased and squashed
